WalletWasabi/Blockchain/Analysis/BlockchainAnalyzer.cs

```diff
@@ -20,58 +21,61 @@ public BlockchainAnalyzer(int privacyLevelThreshold)
 	/// </summary>
 	public void Analyze(SmartTransaction tx)
 	{
-		var inputCount = tx.Transaction.Inputs.Count;
-		var outputCount = tx.Transaction.Outputs.Count;
-
 		var ownInputCount = tx.WalletInputs.Count;
-		var ownOutputCount = tx.WalletOutputs.Count;
+
+		var foreignInputCount = tx.ForeignInputs.Count;
+		var foreignOutputCount = tx.ForeignOutputs.Count;
 
 		if (ownInputCount == 0)
 		{
 			AnalyzeReceive(tx);
 		}
-		else if (inputCount == ownInputCount && outputCount != ownOutputCount)
+		else if (foreignInputCount == 0 && foreignOutputCount > 0)
 		{
 			AnalyzeNormalSpend(tx);
 		}
 		else
 		{
-			AnalyzeWalletInputs(tx, out HashSet<HdPubKey> distinctWalletInputPubKeys, out int newInputAnonset);
+			AnalyzeWalletInputs(tx, out int newInputAnonset, out int sanctionedInputAnonset);
+			AdjustWalletInputsBefore(tx, newInputAnonset);
 
-			if (inputCount == ownInputCount)
+			if (foreignInputCount == 0)
 			{
 				AnalyzeSelfSpend(tx, newInputAnonset);
 			}
 			else
 			{
-				AnalyzeCoinjoin(tx, newInputAnonset, distinctWalletInputPubKeys);
+				AnalyzeCoinjoin(tx, newInputAnonset, sanctionedInputAnonset);
 			}
 
-			AdjustWalletInputs(tx, distinctWalletInputPubKeys, newInputAnonset);
+			AdjustWalletInputsAfter(tx, newInputAnonset);
 		}
 
 		AnalyzeClusters(tx);
 	}
 
 	/// <param name="newInputAnonset">The new anonymity set of the inputs.</param>
-	private void AnalyzeWalletInputs(SmartTransaction tx, out HashSet<HdPubKey> distinctWalletInputPubKeys, out int newInputAnonset)
+	private void AnalyzeWalletInputs(
+		SmartTransaction tx,
+		out int newInputAnonset,
+		out int sanctionedInputAnonset)
 	{
 		// We want to weaken the punishment if the input merge happens in coinjoins.
 		// Our strategy would be is to set the coefficient in proportion to our own inputs compared to the total inputs of the transaction.
 		// However the accuracy can be increased if we consider every input with the same pubkey as a single input entity.
 		// This we can only do for our own inputs as we don't know the pubkeys - nor the scripts - of other inputs.
 		// Another way to think about this is: reusing pubkey on the input side is good, the punishment happened already.
-		distinctWalletInputPubKeys = tx.WalletInputs.Select(x => x.HdPubKey).ToHashSet();
-		var distinctWalletInputPubKeyCount = distinctWalletInputPubKeys.Count;
-		var pubKeyReuseCount = tx.WalletInputs.Count - distinctWalletInputPubKeyCount;
-		double coefficient = (double)distinctWalletInputPubKeyCount / (tx.Transaction.Inputs.Count - pubKeyReuseCount);
+		double coefficient = (double)tx.WalletVirtualInputs.Count / (tx.ForeignInputs.Count + tx.WalletVirtualInputs.Count);
 
-		newInputAnonset = Intersect(distinctWalletInputPubKeys.Select(x => x.AnonymitySet), coefficient);
+		newInputAnonset = Intersect(tx.WalletVirtualInputs.Select(x => x.HdPubKey.AnonymitySet), coefficient);
 
```

Comment:
This deserves a comment. I suggest:
If we remix with siblings from earlier CoinJoins, then we must avoid double-counting their contribution to our anonymity set. Search the history of each input to find any siblings with which we are remixing in the current transaction. Discount the anonymity gain that we got from them earlier and are now losing due to the remix.